### PR TITLE
Fix persistent Welford layer norm numerics

### LINKED ISSUE #186577: `torch.compile` produces wrong results when `F.layer_norm` is fused with `reciprocal`/`rsqrt` epilogue
## Summary

When Inductor fuses `F.layer_norm` with a subsequent division-like pointwise op (`reciprocal`, `rsqrt`, `1/sqrt`), the generated kernel produces significantly wrong results. Each op alone compiles correctly; only the fused combination triggers the bug. `aot_eager` produces correct results, confirming this is an Inductor codegen issue.

## Minimal reproducer

```python
import torch
import torch.nn.functional as F

x = torch.randn(64, 128, device='cuda')

def fn(x):
    return torch.reciprocal(torch.abs(F.layer_norm(x, [128])) + 1e-6)

eager = fn(x)

torch._dynamo.reset()
aot = torch.compile(fn, backend='aot_eager')(x)

torch._dynamo.reset()
compiled = torch.compile(fn, backend='inductor')(x)

print(f"eager vs aot_eager: {(eager - aot).abs().max().item():.4e}")    # 0.0000e+00
print(f"eager vs inductor:  {(eager - compiled).abs().max().item():.4e}")  # ~9e-01
```

## Affected ops (epilogue after `layer_norm`)

| Epilogue op | Triggers? | Max diff |
|-------------|-----------|----------|
| `reciprocal` | Yes | ~80 (shape dependent) |
| `rsqrt` | Yes | ~877 (with cummax prefix) |
| `1/sqrt` (manual) | Yes | ~0.05 |
| `exp` | No | ~7e-6 |
| `tanh` | No | ok |

## Trigger conditions

- **Op pattern**: `F.layer_norm(x, [N])` followed by `reciprocal`, `rsqrt`, or `1/sqrt`
- **Shape dependent**: larger tensors trigger more reliably; small tensors (e.g., `(2, 256)`) may not trigger
- **dtype**: primarily float32; float16 also affected; float64 and bfloat16 appear unaffected
- **Backend**: Inductor only; `aot_eager` is correct
- **Each op alone is correct**: `layer_norm` alone is correct (diff ~7e-7), `reciprocal`/`rsqrt` alone on LN output is correct (diff = 0)
- **Manual normalization does not trigger**: `x / sqrt(mean(x^2) + eps)` + `rsqrt` is fine — only `F.layer_norm` triggers

## Amplified by upstream scan ops

When `cummax`/`cummin`/`cumsum` precedes the pattern, the error is amplified dramatically because the scan output distribution differs from random normal:

```python
x = torch.randn(68, 4, 14, 9, device='cuda')

def fn(x):
    t = torch.cummax(x, dim=-1)[0]         # scan op
    t = F.layer_norm(t, [9])                # layer_norm
    return torch.rsqrt(torch.abs(t) + 1e-6) # rsqrt epilogue

eager = fn(x)
torch._dynamo.reset()
compiled = torch.compile(fn, backend='inductor')(x)
print((eager - compiled).abs().max().item())  # ~877
```

## Shape sensitivity

```
# reciprocal(|LN(x)| + eps), varying shape:
(32, 64)      → diff=0.011
(64, 128)     → diff=0.82
(32, 32, 32)  → diff=81
(8, 4, 14, 9) → diff=0.077
```

## Environment

- PyTorch: 2.13.0.dev20260601+cu126
- GPU: NVIDIA Tesla T4 (CUDA 12.6)
- OS: Linux

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186582

Inductor lowers native layer norm through var_mean. For Triton persistent reductions, welford_reduce was taking the welford_reduce_fallback path, which computes mean and variance with a two-pass formula instead of preserving the Welford accumulation order. The resulting small layer norm differences can be amplified heavily by reciprocal- and rsqrt-like epilogues near zero, producing silent wrong results.

Keep persistent welford_reduce on the Welford lowering path and use masked lane weights to exclude inactive elements. This fixes the root reduction-order mismatch instead of adding a graph-level fusion barrier. Add CUDA regressions for direct layer_norm followed by reciprocal/rsqrt and for the cummax-amplified reciprocal/rsqrt/manual 1/sqrt cases from the issue.

Test Plan:

- python - <<'PY' ... original direct repro with force_disable_caches=True; aot_eager assert_close PASS; inductor assert_close PASS; eager vs inductor max_abs 0.08984375 max_rel 2.673832386790309e-05

- python - <<'PY' ... cummax reciprocal/rsqrt/manual_div repros with force_disable_caches=True; all assert_close PASS; rsqrt max_abs 0.000293731689453125; reciprocal max_abs 0.015625; manual_div max_abs 0.000293731689453125

- python - <<'PY' ... direct unittest shim for CudaReproTests.test_layer_norm_with_reciprocal_like_epilogue; Ran 1 test in 7.883s, OK

- python test/run_test.py --include inductor/test_cuda_repro -- -k test_layer_norm_with_reciprocal_like_epilogue (blocked: Missing pip dependency pytest-rerunfailures)

- lintrunner -a torch/_inductor/codegen/triton.py test/inductor/test_cuda_repro.py (ok No lint issues)

Benchmark Results:

- NVIDIA B200, CUDA event timing, 100 iterations after warmup, force_disable_caches=True

- direct_reciprocal_64x128: before median 0.041136 ms, after median 0.041888 ms

- cummax_rsqrt_68x4x14x9: before median 0.041456 ms, after median 0.041424 ms

Fixes #186577

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by williamwen42** [comment]
@claude are any of the test failures related? Especially `test/inductor/test_torchinductor_strided_blocks.py::TritonBlockPointerTestCPU::test_2d_welford_reduction_size0_expected_num_block_pointers_1_expected_num_triton_kernels_1_expect_fallback_True_cpu`?

**Comment by eellison** [comment]
see comment on original issue
